fix: cert auth method watches cert file change and NewCreds() notific…

jasonjoo2010 opened the pull request:
Hi community and committees,

Recently, our team encountered TLS certificate reloading issue even after adding `reload: true` to `config` when using `cert` method. After checking, it looks like not functioning properly.

Although there was a previous PR(#19002 ) which added certificate reloading capability to auto auth, it doesn't work in the actual use case. After checking further a little bit, I found that the agent blocked at the line [auth.go:526](https://github.com/hashicorp/vault/blob/main/command/agentproxyshared/auth/auth.go#L526) while the server responded with 400 errors at transport layer due to expired certificate. In our case, we use certificates in cloud native applications with 5 days as expiry and refresh the certs every day.

At the same time, the lifetime loop also listens on a channel returned by NewCreds() but it's not implemented in `cert` method. So it may be the best and shortest path adding this support to make the reloading mechanism work.

### Description

So in a nutshell, the `reload` option for `cert` method only guarantee `AuthClient()` return a different client instance when called. It doesn't guarantee the renewed certificate on disk can sooner or later be used in auth. I don't know whether people maintain a good understanding on what this option actually does but it would be better to be more straightforward and more automated in reloading it.

This PR adds implementation of existing `NewCreds()` method in the interface as well as a new optional option `reload_period` for the internal file watching loop. And they only take effect when `reload` is enabled. In common cases, `reload` is the only option people need to set when they need.

### TODO only if you're a HashiCorp employee
- [ ] **Backport Labels:** If this PR is in the ENT repo and needs to be backported, backport  
  to N, N-1, and N-2, using the `backport/ent/x.x.x+ent` labels. If this PR is in the CE repo, you should only backport to N, using the `backport/x.x.x` label, not the enterprise labels.
    - [ ] If this fixes a critical security vulnerability or [severity 1](https://www.hashicorp.com/customer-success/enterprise-support) bug, it will also need to be backported to the current [LTS versions](https://developer.hashicorp.com/vault/docs/enterprise/lts#why-is-there-a-risk-to-updating-to-a-non-lts-vault-enterprise-version) of Vault. To ensure this, use **all** available enterprise labels.
- [ ] **ENT Breakage:** If this PR either 1) removes a public function OR 2) changes the signature
  of a public function, even if that change is in a CE file, _double check_ that
  applying the patch for this PR to the ENT repo and running tests doesn't
  break any tests. Sometimes ENT only tests rely on public functions in CE
  files.
- [ ] **Jira:** If this change has an associated Jira, it's referenced either
  in the PR description, commit message, or branch name.
- [ ] **RFC:** If this change has an associated RFC, please link it in the description.
- [ ] **ENT PR:** If this change has an associated ENT PR, please link it in the
  description. Also, make sure the changelog is in this PR, _not_ in your ENT PR.

VioletHynes commented on command/agentproxyshared/auth/cert/cert.go:
```diff
@@ -87,6 +109,20 @@ func NewCertAuthMethod(conf *auth.AuthConfig) (auth.AuthMethod, error) {
 
+	if c.reload {
+		readPeriod := 1 * time.Minute
+		if jwtReadPeriodRaw, ok := conf.Config["reload_period"]; ok {
```
```suggestion
		if reloadPeriodRaw, ok := conf.Config["reload_period"]; ok {
```

VioletHynes replied:
Seems to be a copy/paste typo :)

jasonjoo2010 replied:
Thanks for pointing out, fixed

VioletHynes commented on command/agentproxyshared/auth/cert/cert.go:
```diff
@@ -87,6 +109,20 @@ func NewCertAuthMethod(conf *auth.AuthConfig) (auth.AuthMethod, error) {
+	if c.reload {
+		readPeriod := 1 * time.Minute
+		if jwtReadPeriodRaw, ok := conf.Config["reload_period"]; ok {
+			jwtReadPeriod, err := parseutil.ParseDurationSecond(jwtReadPeriodRaw)
```
```suggestion
			reloadPeriod, err := parseutil.ParseDurationSecond(reloadPeriodRaw)
```

VioletHynes commented on command/agentproxyshared/auth/cert/cert.go:
```diff
@@ -87,6 +109,20 @@ func NewCertAuthMethod(conf *auth.AuthConfig) (auth.AuthMethod, error) {
+			if err != nil {
+				return nil, fmt.Errorf("error parsing 'jwt_read_period' value: %w", err)
+			}
+			readPeriod = jwtReadPeriod
```
```suggestion
			readPeriod = reloadPeriodRaw
```

VioletHynes commented on command/agentproxyshared/auth/cert/cert.go:
```diff
@@ -87,6 +109,20 @@ func NewCertAuthMethod(conf *auth.AuthConfig) (auth.AuthMethod, error) {
+		if jwtReadPeriodRaw, ok := conf.Config["reload_period"]; ok {
+			jwtReadPeriod, err := parseutil.ParseDurationSecond(jwtReadPeriodRaw)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing 'jwt_read_period' value: %w", err)
```
```suggestion
				return nil, fmt.Errorf("error parsing 'reload_period' value: %w", err)
```

VioletHynes commented on command/agentproxyshared/auth/cert/cert_test.go:
```diff
@@ -188,4 +194,48 @@ func TestCertAuthMethod_AuthClient_withCertsReload(t *testing.T) {
 	}
+
+	method.CredSuccess()
+	ctx, cancel := context.WithTimeout(context.TODO(), 3*time.Second)
```
```suggestion
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
```

VioletHynes commented on command/agentproxyshared/auth/cert/cert_test.go:
```diff
@@ -188,4 +194,48 @@ func TestCertAuthMethod_AuthClient_withCertsReload(t *testing.T) {
+	}
+
+	inst.clientKey = "./test-fixtures/keys/key1.pem"
+	ctx, cancel = context.WithTimeout(context.TODO(), 3*time.Second)
```
```suggestion
	ctx, cancel = context.WithTimeout(context.Background(), 3*time.Second)
```

VioletHynes commented on command/agentproxyshared/auth/cert/cert_test.go:
```diff
@@ -188,4 +194,48 @@ func TestCertAuthMethod_AuthClient_withCertsReload(t *testing.T) {
+	}
+}
+
+func TestHashCert(t *testing.T) {
```
This should have a godoc explaining what the test does and what it's testing

jasonjoo2010 replied:
Hi, I split this unit test into 3 with inline docs and function doc to `hashCert()`. Please take a look if I need to add a more comprehensive on them. Thanks

VioletHynes reviewed (commented):
Hi there! Thanks for your contribution.

Reading through the description and your submission, this makes sense to me. That being said, I'm not an expert in cert logic, so I'm going to try and get a second set of eyes on this that does have that expertise and talk through this with them before I call this approved.

In either case, I've added some comments that will need to be addressed before this is merged. This is shaping up to be awesome so far :D

VioletHynes commented:
To add to the suggestions Scott and I made, this will also need a changelog (a file called `28126.txt` in the `changelog` directory containing a discussion of the improvement) -- you can look at the other files for examples, and I'm happy to review.

You will also want to update [the docs](https://developer.hashicorp.com/vault/docs/agent-and-proxy/autoauth/methods/cert#reload) around this feature. In particular, there should be a new entry added for the new config option, and you should also update the `reload` option as appropriate. This file can be found at `website/content/docs/agent-and-proxy/autoauth/methods/cert.mdx`.

Thanks for the contribution! Excited to get this merged. Please let me know if you have any questions about any of the suggestions.

jasonjoo2010 commented on command/agentproxyshared/auth/cert/cert.go:
```diff
@@ -156,3 +209,73 @@ func (c *certMethod) AuthClient(client *api.Client) (*api.Client, error) {
+	}
+	c.logger.Debug("The cert/key are valid")
+
+	sum := sha1.Sum(append(append(certPEMBlock, keyPEMBlock...), caPEMBlock...))
```
Thanks and no objections, done.

VioletHynes commented on changelog/28126.txt:
````diff
@@ -0,0 +1,4 @@
+```release-note:improvement
+auth/cert: support watching changes on certificate/key files and notifying the auth handler when `enable_reauth_on_new_credentials` is enabled.
+website/docs: add docs for the new option `reload_period` for auth/cert and `enable_reauth_on_new_credentials` for Auto-Auth.
````
```suggestion
```

VioletHynes replied:
No need to add a changelog for docs -- changelogs are for released versions of Vault.

VioletHynes commented on changelog/28126.txt:
````diff
@@ -0,0 +1,4 @@
+```release-note:improvement
+auth/cert: support watching changes on certificate/key files and notifying the auth handler when `enable_reauth_on_new_credentials` is enabled.
````
```suggestion
auto-auth/cert: support watching changes on certificate/key files and notifying the auth handler when `enable_reauth_on_new_credentials` is enabled.
```

VioletHynes commented on command/agentproxyshared/auth/cert/cert_test.go:
```diff
@@ -188,4 +195,104 @@ func TestCertAuthMethod_AuthClient_withCertsReload(t *testing.T) {
+	}
+}
+
+func TestCertAuthMethod_hashCert_withEmptyPaths(t *testing.T) {
```
Sorry I missed this in my last review, but these tests should have GoDocs (a comment above them starting with the test name), e.g.:

```
// TestCertAuthMethod_hashCert_withEmptyPaths is a test that...
```

jasonjoo2010 replied:
I added some doc elaborating the purposes, pls take a look

jasonjoo2010 commented:
Hi Violet and Scott,

After checking the existing code logic carefully, I realised there were special cases that the users may:  

1. Don't give any cert/key files in the config.
2. Only set ca cert.
3. Only set client cert/key.

Although the most common use case is setting all 3 options, we need to cover them in the test cases and keep it backward compatible. So I adjusted the codes a little bit for this.  

I also added the release note and relevant module docs, please take a look at them.  

Any further suggestions, please let me know.

Cheers

VioletHynes commented on website/content/docs/agent-and-proxy/autoauth/index.mdx:
```diff
@@ -105,6 +105,10 @@ The top level `auto_auth` block has two configuration entries:
 
 - `sinks` `(array of objects: optional)` - Configuration for the sinks
 
+- `enable_reauth_on_new_credentials` `(bool: optional)` - If enabled, Auto-auth will
```
In the code, this is documented as the following:
```
	// NOTE: This is unsupported outside of testing and may disappear at any
	// time.
```

As such, I don't feel as though it should be documented. I'm open to hearing if you feel otherwise, though. If we did want to support this and have a strong reason to do so, that would be a different conversation.

VioletHynes replied:
I continued this question in the docs for `reload`.

VioletHynes commented on website/content/docs/agent-and-proxy/autoauth/methods/cert.mdx:
```diff
@@ -31,5 +31,13 @@ config stanza, Agent and Proxy will fall back to using TLS settings from their r
+  Note that `enable_reauth_on_new_credentials` for autoauth needs to be enabled together.
+  See [Auto-Auth Configuration](/vault/docs/agent-and-proxy/autoauth#configuration).
+
+- `reload_period` `(duration: "1m", optional)` - The duration after which the Agent
```
```suggestion
- `reload_period` `(duration: "1m", optional)` - The duration after which auto-auth
```

VioletHynes replied:
auto-auth is supported for both Agent and Proxy, so best to be generic

jasonjoo2010 replied:
Agree, changed it

VioletHynes commented on website/content/docs/agent-and-proxy/autoauth/methods/cert.mdx:
```diff
@@ -31,5 +31,13 @@ config stanza, Agent and Proxy will fall back to using TLS settings from their r
+- `reload` `(bool: optional, default: false)` - If true, causes the local x509
+  key-pair to be reloaded from disk on each authentication attempt. This is useful
+  in situations where client certificates are short-lived and automatically renewed.
+  Note that `enable_reauth_on_new_credentials` for autoauth needs to be enabled together.
```
Could you elaborate more on why `enable_reauth_on_new_credentials` is needed? I'm quite wary of exposing this if we don't need to. I'm also a little confused why this isn't needed in any of the tests that have been added (do we need additional coverage?) and this wasn't mentioned in the description of the PR.

If there's truly no other option for this use case to work than exposing such a config, I'd be open to it, but I'd want to remove the deprecation notice and also add a new changelog for this feature, along with why. First, though -- I'd like to better understand why we need this and what the behaviour would be if this was set to false.

VioletHynes replied:
It would also be important to document what the behaviour would be if this were set to false.

jasonjoo2010 replied:
Hi Violet
Thanks for the comments.
Coming to this "hidden" option, there are also other discussions/requirements from users like this: https://github.com/hashicorp/vault/issues/12233#issuecomment-2016340946

I know there is another way to signal proxy/agent to reload the entire config by using `SIGHUP`, but this is a bit troublesome if the cert provider is another service and it doesn't know who is using that on the node.

However, for `auth/cert`, the `reload` option is not a real reloading feature. Instead, it only guarantees when the main loop in auth handler has a chance for a next loop it reloads the fresh files from disk or just go with a cached HTTP client. And there's no way to do a real reloading without this hidden option which situation is similar to `auth/jwt`.

Certainly, we can leave this option to users who know this trick and keep it remain undocumented.
Wdyt?

VioletHynes replied:
Hmm, from what I can tell in the git blame, this has been there for six years, and we haven't wanted to remove it yet. I think I'm convinced. Let's document it. 

We should delete the two bits of text in Agent/Proxy config that say the following:
```
	// NOTE: This is unsupported outside of testing and may disappear at any
	// time.
```

I'll request a couple more docs updates around this area too, but I'm okay with going ahead with making the formerly undocumented feature real, especially as people seem to want it for other use cases. Thanks for the well-considered response!

jasonjoo2010 replied:
> this has been there for six years

That is really awesome ;)

Do you know the author of it? Not sure if there's any security concern when enabling it. But since it's disabled by default, things are still under control and users can opt-in to it.

VioletHynes replied:
There's no security concern -- users could still use it before, it was just added for testing purposes, it seems.

VioletHynes reviewed (commented):
Hi there! Looking better! I've asked a few clarifying questions and I have a few small suggestions, too.

VioletHynes commented on website/content/docs/agent-and-proxy/autoauth/index.mdx:
```diff
@@ -105,6 +105,10 @@ The top level `auto_auth` block has two configuration entries:
 
 - `sinks` `(array of objects: optional)` - Configuration for the sinks
 
+- `enable_reauth_on_new_credentials` `(bool: optional)` - If enabled, Auto-auth will
+  handle new credential event from supported auth methods(alicloud/aws/cert/jwt/ldap/oci)
+  and re-authenticate with the new credential.
```
```suggestion
- `enable_reauth_on_new_credentials` `(bool: optional)` - If enabled, auto-auth will
  handle new credential events from supported auth methods (AliCloud/AWS/Cert/JWT/LDAP/OCI)
  and re-authenticate with the new credential.
```

VioletHynes commented on website/content/docs/agent-and-proxy/autoauth/methods/cert.mdx:
```diff
@@ -31,5 +31,13 @@ config stanza, Agent and Proxy will fall back to using TLS settings from their r
 
-- `reload` `(bool: optional, default: false)` - If true, causes the local x509 key-pair to be reloaded from disk on each authentication attempt.
-  This is useful in situations where client certificates are short-lived and automatically renewed.
+- `reload` `(bool: optional, default: false)` - If true, causes the local x509
+  key-pair to be reloaded from disk on each authentication attempt. This is useful
+  in situations where client certificates are short-lived and automatically renewed.
+  Note that `enable_reauth_on_new_credentials` for autoauth needs to be enabled together.
+  See [Auto-Auth Configuration](/vault/docs/agent-and-proxy/autoauth#configuration).
```
```suggestion
- `reload` `(bool: optional, default: false)` - If true, causes the local x509
  key-pair to be reloaded from disk on each authentication attempt. This is useful
  in situations where client certificates are short-lived and automatically renewed.
  Note that `enable_reauth_on_new_credentials` for auto-auth will need to be additionally enabled
  for immediate re-auth on a new certificate. See [Auto-Auth Configuration](/vault/docs/agent-and-proxy/autoauth#configuration).
```

VioletHynes reviewed (commented):
This looks good! Just a few small things to get this over the finish line. I agree that we should make `enable_reauth_on_new_credentials` supported.

If you could remove the bits of text:
```
	// NOTE: This is unsupported outside of testing and may disappear at any
	// time.
```
above it (in the two `config.go` files) that'd be great. We should do that if we're documenting it as officially supported.

VioletHynes commented on command/agent/config/config.go:
```diff
@@ -129,8 +129,6 @@ type AutoAuth struct {
 	Method *Method `hcl:"-"`
 	Sinks  []*Sink `hcl:"sinks"`
 
-	// NOTE: This is unsupported outside of testing and may disappear at any
```
Could you also remove the same comment in `command/proxy/config/config.go`? :)

VioletHynes commented on changelog/28126.txt:
````diff
@@ -0,0 +1,3 @@
+```release-note:improvement
+auto-auth/cert: support watching changes on certificate/key files and notifying the auth handler when `enable_reauth_on_new_credentials` is enabled.
+```
````
We should add a new changelog here for adding `enable_reauth_on_new_credentials`. I can't give a great example because GitHub gets confused by the formatting, but you can add two changelog blocks in the same file (see `changelog/24790.txt` as an example.

The second improvement should say something like
`auto-auth: support new config option, `enable_reauth_on_new_credentials`, supporting more frequent authentication on certain auto-auth types`

jasonjoo2010 replied:
Hmmm add a new one in the note, pls have a look if it's in the right format. Although `certain` here is a bit confusing to the user probably

VioletHynes replied:
This looks good, thanks!

VioletHynes commented on website/content/docs/agent-and-proxy/autoauth/methods/cert.mdx:
```diff
@@ -31,5 +31,14 @@ config stanza, Agent and Proxy will fall back to using TLS settings from their r
+  enabled for immediate re-auth on a new certificate.
+  See [Auto-Auth Configuration](/vault/docs/agent-and-proxy/autoauth#configuration).
+
+- `reload_period` `(duration: "1m", optional)` - The duration after which the auto-auth
```
```suggestion
- `reload_period` `(duration: "1m", optional)` - The duration after which auto-auth
```

VioletHynes reviewed (commented):
Looks great! Just a couple tiny things left before I can approve this and get it merged. Thanks for bearing with me throughout all of the comments, and I really appreciate the quick updates and responses.

VioletHynes reviewed (approved):
Thanks for this! Approved. Apologies it took us some time to get to this, and I appreciate your hard work and readiness to update based on feedback.

I've kicked off CI, and assuming it's all green, I'll get this merged :)

Thank you for your contribution!

jasonjoo2010 commented:
I saw there were tiny red ones. No worry, let me solve the race condition in test cases, they are only for debugging purpose but we can do it in the right way.

VioletHynes commented:
Thanks for the contribution, and the diligence in the added and improved tests. 

Merging! 🚢

jasonjoo2010 commented:
Thanks for your kind and patient review, @VioletHynes  and @sgmiller 
That's what makes a product great and greater, really appreciate it!
